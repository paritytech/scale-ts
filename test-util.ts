/// <reference lib="deno.unstable"/>

import { assertEquals, assertThrows } from "https://deno.land/std@0.161.0/testing/asserts.ts"
import { assertSnapshot } from "https://deno.land/std@0.161.0/testing/snapshot.ts"
import { AnyShape, assert, DecodeBuffer, Shape, ShapeAssertError } from "./common/mod.ts"

const [lipsum, words, cargoLock] = ["lipsum.txt", "words.txt", "Cargo.lock"].map((fileName) => () =>
  Deno.readTextFile(fileName)
)
export const files = { lipsum: lipsum!, words: words!, cargoLock: cargoLock! }

export function testShape<T>(
  shape: Shape<T>,
  values: NoInfer<T>[] | Record<string, NoInfer<T> | (() => NoInfer<T>)>,
  async?: boolean,
): void
export function testShape<T>(
  shape: Shape<T>,
  values: T[] | Record<string, T | (() => T)>,
  async = false,
) {
  for (const key in values) {
    let value = values[key as never] as T | (() => T)
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => T)()
      }
      assert(shape, value)
      const encoded = async ? await shape.encodeAsync(value) : shape.encode(value)
      await assertSnapshot(t, encoded, { serializer: serializeU8A })
      const decodeBuffer = new DecodeBuffer(encoded)
      const decoded = shape.subDecode(decodeBuffer)
      assertEquals(decoded, value)
      assertEquals(decodeBuffer.index, encoded.length)
      assert(shape, decoded)
    })
  }
}

export function testInvalid(shape: AnyShape, values: unknown[] | Record<string, unknown | (() => unknown)>) {
  for (const key in values) {
    let value: unknown = values[key as never]
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} invalid ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => unknown)()
      }
      await assertThrowsSnapshot(t, () => assert(shape as Shape<any>, value), ShapeAssertError)
    })
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n")
}

type NoInfer<T> = T extends infer U ? U : never
export function benchShape<T>(name: string, shape: Shape<T>, value: NoInfer<T>): void
export function benchShape<T>(name: string, shape: Shape<T>, value: T) {
  const encoded = shape.encode(value)
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    shape.encode(value)
  })
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    shape.decode(encoded)
  })
}

export async function assertThrowsSnapshot(
  t: Deno.TestContext,
  fn: () => unknown,
  ctor: abstract new(...args: any) => Error = Error,
) {
  try {
    fn()
  } catch (error) {
    if (!(error instanceof ctor)) {
      throw new Error("Expected " + ctor.name)
    }
    return await assertSnapshot(t, `${error.name}: ${error.message}`, { serializer: (x) => x })
  }
  throw new Error("Expected function to throw")
}

export { assertEquals, assertSnapshot, assertThrows }
