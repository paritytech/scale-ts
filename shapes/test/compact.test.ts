import * as $ from "../../mod.ts"
import { testInvalid, testShape } from "../../test-util.ts"

testShape($.compact($.u32), [0, 1, 6, 8, 14, 16, 30, 32].map((x) => (2 ** x) - 1))
testShape($.compact($.u256), [0, 6, 14, 30, 40, 64, 128, 128, 256].map((x) => (1n << BigInt(x)) - 1n))

testShape($.compact($.tuple($.u32)), [[123]])
testShape($.compact($.field("foo", $.u32)), [{ foo: 456 }])
testShape($.compact($.object($.field("foo", $.u32))), [{ foo: 456 }])

testInvalid($.compact($.u8), [
  null,
  undefined,
  -1,
  256,
])

testInvalid($.compact($.u128), [
  null,
  undefined,
  -1,
  100000000000000000000000000000000000000000000000000000000n,
])

testInvalid($.compact($.tuple($.u32)), [
  null,
  undefined,
  [],
  [-1],
  [123n],
  [Number.MAX_SAFE_INTEGER],
])

testInvalid($.compact($.field("foo", $.u32)), [
  null,
  undefined,
  {},
  { foo: -1 },
  { foo: 123n },
  { foo: Number.MAX_SAFE_INTEGER },
])
