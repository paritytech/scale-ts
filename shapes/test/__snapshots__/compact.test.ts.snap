export const snapshot = {};

snapshot[`\$.compact(\$.u32) 0 1`] = `00`;

snapshot[`\$.compact(\$.u32) 1 1`] = `04`;

snapshot[`\$.compact(\$.u32) 63 1`] = `fc`;

snapshot[`\$.compact(\$.u32) 255 1`] = `
fd
03
`;

snapshot[`\$.compact(\$.u32) 16383 1`] = `
fd
ff
`;

snapshot[`\$.compact(\$.u32) 65535 1`] = `
fe
ff
03
00
`;

snapshot[`\$.compact(\$.u32) 1073741823 1`] = `
fe
ff
ff
ff
`;

snapshot[`\$.compact(\$.u32) 4294967295 1`] = `
03
ff
ff
ff
ff
`;

snapshot[`\$.compact(\$.u256) 0n 1`] = `00`;

snapshot[`\$.compact(\$.u256) 63n 1`] = `fc`;

snapshot[`\$.compact(\$.u256) 16383n 1`] = `
fd
ff
`;

snapshot[`\$.compact(\$.u256) 1073741823n 1`] = `
fe
ff
ff
ff
`;

snapshot[`\$.compact(\$.u256) 1099511627775n 1`] = `
07
ff
ff
ff
ff
ff
`;

snapshot[`\$.compact(\$.u256) 18446744073709551615n 1`] = `
13
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.compact(\$.u256) 340282366920938463463374607431768211455n 1`] = `
33
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.compact(\$.u256) 340282366920938463463374607431768211455n 2`] = `
33
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.compact(\$.u256) 115792089237316195423570985008687907853269984665640564039457584007913129639935n 1`] = `
73
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.compact(\$.tuple(\$.u32)) [ 123 ] 1`] = `
ed
01
`;

snapshot[`\$.compact(\$.field("foo", \$.u32)) { foo: 456 } 1`] = `
21
07
`;

snapshot[`\$.compact(\$.object(\$.field("foo", \$.u32))) { foo: 456 } 1`] = `
21
07
`;

snapshot[`\$.compact(\$.u8) invalid null 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.compact(\$.u8) invalid undefined 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.compact(\$.u8) invalid -1 1`] = `ShapeAssertError: value < 0`;

snapshot[`\$.compact(\$.u8) invalid 256 1`] = `ShapeAssertError: value > 255`;

snapshot[`\$.compact(\$.u128) invalid null 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.compact(\$.u128) invalid undefined 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.compact(\$.u128) invalid -1 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.compact(\$.u128) invalid 100000000000000000000000000000000000000000000000000000000n 1`] = `ShapeAssertError: value > 340282366920938463463374607431768211455n`;

snapshot[`\$.compact(\$.tuple(\$.u32)) invalid null 1`] = `ShapeAssertError: !(value instanceof Array)`;

snapshot[`\$.compact(\$.tuple(\$.u32)) invalid undefined 1`] = `ShapeAssertError: !(value instanceof Array)`;

snapshot[`\$.compact(\$.tuple(\$.u32)) invalid [] 1`] = `ShapeAssertError: value.length !== 1`;

snapshot[`\$.compact(\$.tuple(\$.u32)) invalid [ -1 ] 1`] = `ShapeAssertError: value[0] < 0`;

snapshot[`\$.compact(\$.tuple(\$.u32)) invalid [ 123n ] 1`] = `ShapeAssertError: typeof value[0] !== "number"`;

snapshot[`\$.compact(\$.tuple(\$.u32)) invalid [ 9007199254740991 ] 1`] = `ShapeAssertError: value[0]: invalid int`;

snapshot[`\$.compact(\$.field("foo", \$.u32)) invalid null 1`] = `ShapeAssertError: value == null`;

snapshot[`\$.compact(\$.field("foo", \$.u32)) invalid undefined 1`] = `ShapeAssertError: typeof value !== "object"`;

snapshot[`\$.compact(\$.field("foo", \$.u32)) invalid {} 1`] = `ShapeAssertError: !("foo" in value)`;

snapshot[`\$.compact(\$.field("foo", \$.u32)) invalid { foo: -1 } 1`] = `ShapeAssertError: value.foo < 0`;

snapshot[`\$.compact(\$.field("foo", \$.u32)) invalid { foo: 123n } 1`] = `ShapeAssertError: typeof value.foo !== "number"`;

snapshot[`\$.compact(\$.field("foo", \$.u32)) invalid { foo: 9007199254740991 } 1`] = `ShapeAssertError: value.foo: invalid int`;
